src/server.cpp

```diff
@@ -593,6 +595,26 @@ void Server::AsyncRunStep(bool initial_step)
 	}
 
 	/*
+		Listen to the admin chat, if available
+	*/
+	if (m_admin_chat) {
+		MutexAutoLock lock(m_env_mutex);
```

Comment:
Well, we don't know here whether envlock needs to be set, so we assume we need it for `handleAdminChat`. Now I could lock the environment inside `handleAdminChat` as instead, but then I would have to modify how the environment is locked for the regular chat, where the environment is locked in `Server::ProcessData`. Removing it from there isn't preferrable. Now I could move the locking of the mutex into the while loop, or even into the if case. It wouldn't make much sense however, as 99% of the events received from the admin chat interface are chat messages (the other type can be nick informs, but they only happen at startup). One optimisation I could do is to move the lock into an if case that checks whether there are elements in the queue at all.